tests/jobs/test_scheduler_job.py

```diff
@@ -2367,6 +2370,194 @@ def test_runs_are_created_after_max_active_runs_was_reached(self, dag_maker, ses
+        ti.span_status = SpanStatus.SHOULD_END
+        ti.context_carrier = {}
+        session.merge(ti)
+        session.merge(dr)
+        session.commit()
+
+        dr_span = Trace.start_root_span(span_name="dag_run_span", start_as_current=False)
+        ti_span = Trace.start_child_span(span_name="ti_span", start_as_current=False)
+
+        self.job_runner.active_spans.set(dr.run_id, dr_span)
+        self.job_runner.active_spans.set(ti.key, ti_span)
+
+        # Given
+        assert dr.span_status == SpanStatus.SHOULD_END
+        assert ti.span_status == SpanStatus.SHOULD_END
+
+        assert self.job_runner.active_spans.get(dr.run_id) is not None
+        assert self.job_runner.active_spans.get(ti.key) is not None
+
+        # When
+        self.job_runner._end_spans_of_externally_ended_ops(session)
+
+        # Then
+        assert dr.span_status == SpanStatus.ENDED
+        assert ti.span_status == SpanStatus.ENDED
+
+        assert self.job_runner.active_spans.get(dr.run_id) is None
+        assert self.job_runner.active_spans.get(ti.key) is None
+
+    @pytest.mark.parametrize(
+        "state, final_span_status",
+        [(State.SUCCESS, SpanStatus.ENDED), (State.RUNNING, SpanStatus.NEEDS_CONTINUANCE)],
+    )
+    def test_end_active_spans(self, state, final_span_status, dag_maker):
+        with dag_maker(
+            dag_id="test_end_active_spans",
+            start_date=DEFAULT_DATE,
+            max_active_runs=1,
+            dagrun_timeout=datetime.timedelta(seconds=60),
+        ):
+            EmptyOperator(task_id="dummy")
+
+        session = settings.Session()
+
+        job = Job()
+        job.id = 1
+        job.job_type = SchedulerJobRunner.job_type
+
+        self.job_runner = SchedulerJobRunner(job=job)
+        self.job_runner.active_spans = ThreadSafeDict()
+        assert len(self.job_runner.active_spans.get_all()) == 0
+
+        dr = dag_maker.create_dagrun(external_trigger=True)
+        dr.state = state
+        dr.span_status = SpanStatus.ACTIVE
+
+        ti = dr.get_task_instances(session=session)[0]
+        ti.state = state
+        ti.span_status = SpanStatus.ACTIVE
+        ti.context_carrier = {}
+        session.merge(ti)
+        session.merge(dr)
+        session.commit()
+
+        dr_span = Trace.start_root_span(span_name="dag_run_span", start_as_current=False)
+        ti_span = Trace.start_child_span(span_name="ti_span", start_as_current=False)
+
+        self.job_runner.active_spans.set(dr.run_id, dr_span)
+        self.job_runner.active_spans.set(ti.key, ti_span)
+
+        # Given
+        assert dr.span_status == SpanStatus.ACTIVE
+        assert ti.span_status == SpanStatus.ACTIVE
+
+        assert self.job_runner.active_spans.get(dr.run_id) is not None
+        assert self.job_runner.active_spans.get(ti.key) is not None
+        assert len(self.job_runner.active_spans.get_all()) == 2
+
+        # When
+        self.job_runner._end_active_spans(session)
+
+        # Then
```

Comment:
Purely up to you, in general we don't use the GIVEN/WHEN/THEN comments, but you can leave them if you like them.

Reply:
I added them for clarity but it's best to follow the project practices. I removed them.

Reply:
I totally get it, I used to put them in as well but dropped the habit to match the Airflow style.  To be fair, now that I am used to it, a blank line really does do pretty much the same thing, so it's not a big deal.